### PR TITLE
pageserver: always load remote metadata (no more `spawn_load`)

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

The pageserver had two ways of loading a tenant:
- `spawn_load` would trust on-disk content to reflect all existing timelines
- `spawn_attach` would list timelines in remote storage.

It was incorrect for `spawn_load` to trust local disk content, because it doesn't know if the tenant might have been attached and written somewhere else.  To make this correct would requires some generation number checks, but the payoff is to avoid one S3 op per tenant at startup, so it's not worth the complexity -- it is much simpler to have one way to load a tenant.

## Summary of changes

- `Tenant` objects are always created with `Tenant::spawn`: there is no more distinction between "load" and  "attach".  
- The ability to run without remote storage (for `neon_local`) is preserved by adding a branch inside `attach` that uses a fallback `load_local` if no remote_storage is present.
- Fix attaching a tenant when it has a timeline with no IndexPart: this can occur if a newly created timeline manages to upload a layer before it has uploaded an index.
- The attach marker file that used to indicate whether a tenant should be "loaded" or "attached" is no longer needed, and is removed.
- The GenericRemoteStorage interface gets a `list()` method that maps more directly to what ListObjects does, returning both keys and common prefixes.  The existing `list_files` and `list_prefixes` methods are just calls into `list()` now -- these can be removed later if we would like to shrink the interface a bit.
- The remote deletion marker is moved into `timelines/` and detected as part of listing timelines rather than as a separate GET request.  If any existing tenants have a marker in the old location (unlikely, only happens if something crashes mid-delete), then they will rely on the control plane retrying to complete their deletion.
- Revise S3 calls for timeline listing and tenant load to take a cancellation token, and retry forever: it never makes sense to make a Tenant broken because of a transient S3 issue.

## Breaking changes

- The remote deletion marker is moved from `deleted` to `timelines/deleted` within the tenant prefix.  Markers in the old location will be ignored: it is the control plane's responsibility to retry deletions until they succeed.  Markers in the new location will be tolerated by the previous release of pageserver via https://github.com/neondatabase/neon/pull/5632
- The local `attaching` marker file is no longer written.  Therefore, if the pageserver is downgraded after running this code, the old pageserver will not be able to distinguish between partially attached tenants and fully attached tenants.  This would only impact tenants that were partway through attaching at the moment of downgrade.  In the unlikely even t that we do experience an incident that prompts us to roll back, then we may check for attach operations in flight, and manually insert `attaching` marker files as needed.

### CODE REVIEW

**Comment by jcsp** [comment]
Testing at 10k tenant scale on i3en.3xlarge:
- Functionally no issues.
- The time to `initial_tenant_load_remote` is about 10 seconds.  Comparing with `main` code in the same environment there is no regression.
- One can persuade the remote load to go faster by increasing concurrency and delaying the rest of tenants starting up until they've all loaded remote content, but the improvement is only about 30%, and one gets hundreds of SlowDowns in the log, which is harmless but annoying.  If we want to make it much faster in future we'll probably want to introduce tenant metadata files we can GET, but for now the simplicity of just listing is nice.

**Comment by jcsp** [comment]
Failure  in `test_pageserver_chaos` looks like something unrelated to this PR (https://github.com/neondatabase/neon/issues/5404#issuecomment-1771630164)

**Comment by problame** [comment]
> The ability to run without remote storage (for neon_local) is preserved by adding a branch inside attach that uses a fallback load_local if no remote_storage is present.

Sigh, one of the remaining items of #5086 was to get rid of the `remote_storage = None` configuration. Hope this wasn't too much pain.

**Comment by jcsp** [comment]
>Sigh, one of the remaining items of https://github.com/neondatabase/neon/pull/5086 was to get rid of the remote_storage = None configuration. Hope this wasn't too much pain.

This was just a case of keeping a little of the old code: we may be  able to just change the defaults for neon_local to run with local_fs and then get rid of it entirely.

File: test_runner/fixtures/neon_fixtures.py [hunk 1]
```diff
@@ -1680,6 +1681,41 @@ def stop(self, immediate: bool = False) -> "NeonPageserver":
...
+        """
+
+        stable_states = {"Active", "Broken"}
+
+        client = self.http_client()
+
+        def complete():
+            log.info("Checking tenants...")
+            tenants = client.tenant_list()
+            tenants = client.tenant_list()
```
Comment (shanyp) [inline test_runner/fixtures/neon_fixtures.py:1708]: duplicated line

**Comment by shanyp** [comment]
> Fix attaching a tenant when it has a timeline with no IndexPart: this can occur if a newly created timeline manages to upload a layer before it has uploaded an index.

Why not make this serial ? first upload the indexPart and only than upload the layer ?

**Comment by shanyp** [comment]
> Revise S3 calls for timeline listing and tenant load to take a cancellation token, and retry forever: it never makes sense to make a Tenant broken because of a transient S3 issue.

While this takes place (retries ..) the PS doesn't start listen to incoming requests, right ?

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -786,6 +841,69 @@ impl Tenant {
         Ok(())
     }
 
+    /// Check for any local timeline directories that are temporary, or do not correspond to a
+    /// timeline that still exists: this can happen if we crashed during a deletion/creation, or
+    /// if a timeline was deleted while the tenant was attached to a different pageserver.
+    fn clean_up_timelines(&self, existent_timelines: &HashSet<TimelineId>) -> anyhow::Result<()> {
```
Comment (problame) [inline pageserver/src/tenant.rs:847]: More of an observation that should be reflected in the PR description & PR label: this is a breaking change.
Once we deploy code with this PR, there is no way to roll back because we don't write the attaching marker anymore => old code will treat incomplete attachment (:= interrupted load_remote_timelines) as completely attached and may start GC'ing data because it doesn't know about some timelines.

File: pageserver/src/tenant.rs [hunk 3]
```diff
@@ -772,12 +832,7 @@ impl Tenant {
             .map_err(LoadLocalTimelineError::ResumeDeletion)?;
         }
 
-        if let AttachMarkerMode::Expect = expect_marker {
-            std::fs::remove_file(&marker_file)
-                .with_context(|| format!("unlink attach marker file {marker_file}"))?;
-            crashsafe::fsync(marker_file.parent().expect("marker file has parent dir"))
-                .context("fsync tenant directory after unlinking attach marker file")?;
-        }
+        self.clean_up_timelines(&existent_timelines)?;
```
Comment (problame) [inline pageserver/src/tenant.rs:835]:
```suggestion
        // The local filesystem contents are a cache of what's in the remote IndexPart;
        // IndexPart is the source of truth.
        self.clean_up_timelines(&existent_timelines)?;
```

File: pageserver/src/tenant/remote_timeline_client/download.rs [hunk 4]
```diff
@@ -170,58 +170,64 @@ pub fn is_temp_download_file(path: &Utf8Path) -> bool {
...
+    let mut other_prefixes = HashSet::new();
+
+    tracing::info!("list_remote_timelines prefixes:");
+    for p in &listing.prefixes {
+        tracing::info!("  '{p}'");
+    }
+    tracing::info!("list_remote_timelines keys:");
+    for p in &listing.keys {
+        tracing::info!("  '{p}'");
+    }
```
Comment (problame) [inline pageserver/src/tenant/remote_timeline_client/download.rs:198]: This will be a storm of messages every startup. I suggest to log this at `debug!` level.
A summary log message at info! level might be ok.

Also, please use `#[instrument]` on the `list_remote_timelines` (not part of the suggestion!) fn and use structured logging.

```suggestion
    for prefix in &listing.prefixes {
        tracing::debug!(prefix, "found prefix");
    }
    for key in &listing.keys {
        tracing::debug!(key, "found key");
    }
    tracing::info!(prefix_count=listing.prefixes.len(), key_count=listings.keys.len(), "summary");
```

File: libs/remote_storage/src/lib.rs [hunk 5]
```diff
@@ -154,7 +175,21 @@ pub trait RemoteStorage: Send + Sync + 'static {
...
+    async fn list(
+        &self,
+        prefix: Option<&RemotePath>,
+        _mode: ListingMode,
+    ) -> anyhow::Result<Listing, DownloadError>; /* {
+                                                     // XXX Placeholder impl.
+                                                     let mut result = Listing::default();
+                                                     result.prefixes = self.list_prefixes(prefix).await?;
+                                                     Ok(result)
+                                                 }*/
```
Comment (problame) [inline libs/remote_storage/src/lib.rs:192]: At a minimum, please format the XXX comment more nicely.

Also, I would love some basic tests for this that are run for each impl of this interface.

File: pageserver/src/tenant/delete.rs [hunk 6]
```diff
@@ -60,7 +58,7 @@ fn remote_tenant_delete_mark_path(
         .context("Failed to strip workdir prefix")
         .and_then(RemotePath::new)
         .context("tenant path")?;
-    Ok(tenant_remote_path.join(Utf8Path::new("deleted")))
+    Ok(tenant_remote_path.join(Utf8Path::new("timelines/deleted")))
```
Comment (problame) [inline pageserver/src/tenant/delete.rs:61]: Where do we do the migration for tenants that already have the old `deleted` mark? 
Is it safe to not do a migration? I haven't though that through, please explain.

How about calling the mark `deleted_v2` then to make it a bit clearer that there was a mark at a previous location?
Prehaps leave a comment about where the previous location was

Also, as with the local state changes: we should acknowledge this is a breaking change, we can't roll back the software because the old software won't know about the new location (and I assume we haven't thought through what happens if the old software encounters a `timelines/deleted` key).

File: pageserver/src/tenant/mgr.rs [hunk 7]
```diff
@@ -660,13 +634,13 @@ pub(crate) async fn create_tenant(
...
+        super::create_tenant_files(conf, &location_conf, &tenant_id).await?;
+
         // TODO: tenant directory remains on disk if we bail out from here on.
         //       See https://github.com/neondatabase/neon/issues/4233
 
-        let created_tenant =
-            schedule_local_tenant_processing(conf, tenant_id, &tenant_directory,
-                AttachedTenantConf::try_from(location_conf)?, resources, None, &TENANTS, ctx)?;
+        let created_tenant = Tenant::spawn(conf, tenant_id, resources,
+        AttachedTenantConf::try_from(location_conf)?, None, &TENANTS, SpawnMode::Create, ctx)?;
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:643]: Turns out a long string was preventing cargo fmt from doing its thing, fixed.  107c643ab7e0b0e67b3b5c6ed3d080b73fd82862

File: pageserver/src/tenant/mgr.rs [hunk 8]
```diff
@@ -660,13 +634,13 @@ pub(crate) async fn create_tenant(
...
-        let tenant_directory = super::create_tenant_files(conf, &location_conf, &tenant_id, CreateTenantFilesMode::Create).await?;
+        super::create_tenant_files(conf, &location_conf, &tenant_id).await?;
+
         // TODO: tenant directory remains on disk if we bail out from here on.
         //       See https://github.com/neondatabase/neon/issues/4233
 
-        let created_tenant =
-            schedule_local_tenant_processing(conf, tenant_id, &tenant_directory,
-                AttachedTenantConf::try_from(location_conf)?, resources, None, &TENANTS, ctx)?;
+        let created_tenant = Tenant::spawn(conf, tenant_id, resources,
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:642]: I think the end state will be that we call directly through to Tenant::spawn without needing this wrapper -- it's currently doing wasteful I/O by having `schedule_local_tenant_processing` stat() tenant directory, where in some call sites we _just_ created that dir.

But yeah, this is a bit of an in-between state, I've cleaned it up by renaming `schedule_local_tenant_processing` to `tenant_spawn` to reflect that it's just a wrapper now. d092cb804

File: pageserver/src/tenant.rs [hunk 9]
```diff
@@ -629,96 +687,98 @@ impl Tenant {
...
+            self.tenant_id,
+            cancel.clone(),
+        )
+        .await?;
+        info!("found {} timelines", remote_timeline_ids.len());
+
+        let deleting = other_keys.contains(TENANT_DELETED_MARKER_FILE_NAME);
+        for k in other_keys {
+            info!("found non timeline key {k}");
+        }
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant.rs:709]: fe61e8f65

File: pageserver/src/tenant.rs [hunk 10]
```diff
@@ -609,7 +667,7 @@ impl Tenant {
                     }
                 }
 
-                match tenant_clone.attach(&ctx, expect_marker).await {
+                match tenant_clone.attach(init_order, preload, &ctx).await {
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant.rs:670]: Right.

File: pageserver/src/tenant/remote_timeline_client/download.rs (hunk 4 above)
Comment (jcsp) [reply pageserver/src/tenant/remote_timeline_client/download.rs:198]: Oops, this was debug code that I didn't mean to commit.  I've just removed it.

File: libs/remote_storage/src/lib.rs (hunk 5 above)
Comment (jcsp) [reply libs/remote_storage/src/lib.rs:192]: Oops, this was commented out while I was refactoring.  I've just removed it now.

Comment (jcsp) [reply libs/remote_storage/src/lib.rs:192]: In 74d9fcde6888191de29a083daffd2a1cc66e1048 I've added a basic test for LocalFS, just so that we have something that tracks the expected behavior.  I don't know that we have infrastructure for unit testing Azure+S3 code: seems like we rely on the coverage we get in integration tests.

File: pageserver/src/tenant.rs [hunk 11]
```diff
@@ -629,96 +687,98 @@ impl Tenant {
...
         let mut timeline_ancestors = HashMap::new();
-        while let Some(result) = part_downloads.join_next().await {
-            // NB: we already added timeline_id as context to the error
-            let result: Result<_, anyhow::Error> = result.context("joinset task join")?;
-            let (timeline_id, client, index_part) = result?;
-            debug!("successfully downloaded index part for timeline {timeline_id}");
+        let mut existent_timelines = HashSet::new();
+        for (timeline_id, preload) in preload.timelines {
+            // In this context a timeline "exists" if it has any content in remote storage: this will
+            // be our cue to not delete any corresponding local directory
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant.rs:763]: Removed in f778bf3b0

File: pageserver/src/tenant.rs (hunk 2 above)
Comment (jcsp) [reply pageserver/src/tenant.rs:856]: Will pick this up in the other comment about forward compat below.

File: pageserver/src/tenant/delete.rs (hunk 6 above)
Comment (jcsp) [reply pageserver/src/tenant/delete.rs:61]: > Where do we do the migration for tenants that already have the old deleted mark?
> Is it safe to not do a migration? I haven't though that through, please explain.

The note in the PR comment is:
```
- The remote deletion marker is moved into `timelines/` and detected as part of listing timelines rather than as a separate GET request.  If any existing tenants have a marker in the old location (unlikely, only happens if something crashes mid-delete), then they will rely on the control plane retrying to complete their deletion.
```

> How about calling the mark deleted_v2 then to make it a bit clearer that there was a mark at a previous location?

Hmm, I think we disambiguating these via the different path is sufficient -- if we call it v2 it's liable to make the code just a bit less readable in the long term as people wonder what v1 was.

> Also, as with the local state changes: we should acknowledge this is a breaking change, we can't roll back the software because the old software won't know about the new location (and I assume we haven't thought through what happens if the old software encounters a timelines/deleted key).

I think it's worth doing a quick PR to let the old version understand the new deletion marker so that we aren't totally incompatible.  We'll still have the rollback issue that local state will be interpreted as authoritative due to the lack of an attach marker, but that will be limited to any tenants that were in a transient state at the time we rolled back, and these can be fixed with a detach/attach cycle.

Comment (jcsp) [reply pageserver/src/tenant/delete.rs:61]: I've opened https://github.com/neondatabase/neon/pull/5632 to give us a realistic possibility of rolling back if we have issues with this.

**Comment by jcsp** [comment]
>> Fix attaching a tenant when it has a timeline with no IndexPart: this can occur if a newly created timeline manages to upload a layer before it has uploaded an index.

>Why not make this serial ? first upload the indexPart and only than upload the layer ?

The correctness rules for indexparts are that layers are uploaded before indexpart references them, and they are removed from indexpart before the layer is deleted.  That way the IndexPart always refers to objects that exist in S3.

**Comment by jcsp** [comment]
>> Revise S3 calls for timeline listing and tenant load to take a cancellation token, and retry forever: it never makes sense to make a Tenant broken because of a transient S3 issue.
> While this takes place (retries ..) the PS doesn't start listen to incoming requests, right ?

The way I read the code in `pageserver.rs`, I think we actually _do_ start listening to incoming requests before tenant loading is complete.  Which is perhaps a problem in its own right, but this PR shouldn't disrupt anything there.

File: pageserver/src/tenant/mgr.rs [hunk 12]
```diff
@@ -1101,17 +1073,10 @@ pub(crate) async fn attach_tenant(
 ) -> Result<(), TenantMapInsertError> {
     tenant_map_insert(tenant_id, || async {
         let location_conf = LocationConf::attached_single(tenant_conf, generation);
-        let tenant_dir = create_tenant_files(conf, &location_conf, &tenant_id, CreateTenantFilesMode::Attach).await?;
+        let tenant_dir = create_tenant_files(conf, &location_conf, &tenant_id).await?;
```
[the remark this answers is not in the document]
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs:1106]: https://github.com/neondatabase/neon/issues/5652

File: pageserver/src/tenant/delete.rs (hunk 6 above)
Comment (jcsp) [reply pageserver/src/tenant/delete.rs:61]: #5632 merged.

**Review by problame — APPROVED** [review]
Couldn't spend much more time reviewing, but, this is definitely moving pageserver in the right direction.

We agreed offline that it's ok to take this breaking change.

I added the right "red label" to the PR, in case people still pay attention to this.

@jcsp  please create a calendar entry for you to shepherd this into prod on next week's release. This means particularly shutting down any ideas about rollback.

File: pageserver/src/tenant/delete.rs [hunk 13]
```diff
@@ -378,7 +349,7 @@ impl DeleteTenantFlow {
 
     pub(crate) async fn should_resume_deletion(
         conf: &'static PageServerConf,
-        remote_storage: Option<&GenericRemoteStorage>,
+        remote_mark_exists: bool,
         tenant: &Tenant,
     ) -> Result<Option<DeletionGuard>, DeleteTenantError> {
```
Comment (koivunej) [inline pageserver/src/tenant/delete.rs:354]: drive-by: this is infallible.

File: pageserver/src/tenant.rs [hunk 14]
```diff
@@ -785,6 +849,69 @@ impl Tenant {
...
+            }
+        };
+
+        for entry in entries {
+            let entry = entry.context("read timeline dir entry")?;
+            let entry_path = entry.path();
+
+            let purge = if crate::is_temporary(entry_path)
+                // TODO: uninit_mark isn't needed any more, since uninitialized timelines are already
+                // covered by the check that the timeline must exist in remote storage.
```
Comment (koivunej) [inline pageserver/src/tenant.rs:875]: this isn't true: see #investigation-2023-11-23-stuck-tenant-deletion
